File: iOS/Inkwell/SharedModelMappers.swift
//
//  SharedModelMappers.swift
//  Inkwell
//
//  Bridges between iOS native model types and the neutral shared KMP model types.
//  Networking and serialization remain native; only the data shapes are shared.
//

import Foundation
import ATProtoKit
import InkwellShared

// ── BlobRef ────────────────────────────────────────────────────────────────

extension ComAtprotoLexicon.Repository.UploadBlobOutput {
    func toShared() -> BlobRef {
        BlobRef(
            link: self.reference.link,
            size: Int32(self.size),
            type: self.type ?? "blob",
            mimeType: self.mimeType
        )
    }
}

extension BlobRef {
    func toiOS() -> ComAtprotoLexicon.Repository.UploadBlobOutput {
        ComAtprotoLexicon.Repository.UploadBlobOutput(
            type: self.type,
            reference: ComAtprotoLexicon.Repository.BlobReference(link: self.link),
            mimeType: self.mimeType ?? "",
            size: Int(self.size)
        )
    }
}

// ── StrongRef ──────────────────────────────────────────────────────────────

extension ComAtprotoLexicon.Repository.StrongReference {
    func toShared() -> StrongRef {
        StrongRef(
            uri: self.recordURI,
            cid: self.recordCID
        )
    }
}

extension StrongRef {
    func toiOS() -> ComAtprotoLexicon.Repository.StrongReference {
        ComAtprotoLexicon.Repository.StrongReference(
            recordURI: self.uri,
            cidHash: self.cid ?? ""
        )
    }
}

// ── Theme Types ────────────────────────────────────────────────────────────

extension SiteStandardLexicon.Theme.RGBColor {
    func toShared() -> RgbColor {
        RgbColor(
            type: Self.type,
            r: Int32(self.r),
            g: Int32(self.g),
            b: Int32(self.b)
        )
    }
}

extension RgbColor {
    func toiOS() -> SiteStandardLexicon.Theme.RGBColor {
        SiteStandardLexicon.Theme.RGBColor(
            r: Int(self.r),
            g: Int(self.g),
            b: Int(self.b)
        )
    }
}

extension SiteStandardLexicon.Theme.RGBAColor {
    func toShared() -> RgbaColor {
        RgbaColor(
            type: Self.type,
            r: Int32(self.r),
            g: Int32(self.g),
            b: Int32(self.b),
            a: Int32(self.a)
        )
    }
}

extension RgbaColor {
    func toiOS() -> SiteStandardLexicon.Theme.RGBAColor {
        SiteStandardLexicon.Theme.RGBAColor(
            r: Int(self.r),
            g: Int(self.g),
            b: Int(self.b),
            a: Int(self.a)
        )
    }
}

extension SiteStandardLexicon.Theme.ColorValue {
    func toShared() -> ColorValue {
        ColorValue(
            type: self.type ?? "",
            r: Int32(self.r),
            g: Int32(self.g),
            b: Int32(self.b),
            a: self.a.map { KotlinInt(value: Int32($0)) }
        )
    }
}

extension ColorValue {
    func toiOS() -> SiteStandardLexicon.Theme.ColorValue {
        SiteStandardLexicon.Theme.ColorValue(
            type: self.type,
            r: Int(self.r),
            g: Int(self.g),
            b: Int(self.b),
            a: self.a.map { Int($0.intValue) }
        )
    }
}

extension SiteStandardLexicon.Theme.LegacyPalette {
    func toShared() -> LegacyPalette {
        LegacyPalette(
            background: self.background,
            text: self.text,
            accent: self.accent,
            link: self.link,
            surfaceHover: self.surfaceHover
        )
    }
}

extension LegacyPalette {
    func toiOS() -> SiteStandardLexicon.Theme.LegacyPalette {
        SiteStandardLexicon.Theme.LegacyPalette(
            background: self.background,
            text: self.text,
            accent: self.accent,
            link: self.link,
            surfaceHover: self.surfaceHover
        )
    }
}

extension SiteStandardLexicon.Theme.BasicDefinition {
    func toShared() -> BasicTheme {
        BasicTheme(
            type: Self.type,
            background: self.background.toShared(),
            foreground: self.foreground.toShared(),
            accent: self.accent.toShared(),
            accentForeground: self.accentForeground.toShared()
        )
    }
}

extension BasicTheme {
    func toiOS() -> SiteStandardLexicon.Theme.BasicDefinition {
        SiteStandardLexicon.Theme.BasicDefinition(
            background: self.background.toiOS(),
            foreground: self.foreground.toiOS(),
            accent: self.accent.toiOS(),
            accentForeground: self.accentForeground.toiOS()
        )
    }
}

extension SiteStandardLexicon.Theme.PublicationTheme {
    func toShared() -> PublicationTheme {
        PublicationTheme(
            type: self.type ?? "",
            backgroundColor: self.backgroundColor?.toShared(),
            pageBackground: self.pageBackground?.toShared(),
            primary: self.primary?.toShared(),
            accentBackground: self.accentBackground?.toShared(),
            accentText: self.accentText?.toShared(),
            pageWidth: self.pageWidth.map { KotlinInt(value: Int32($0)) },
            showPageBackground: self.showPageBackground.map { KotlinBoolean(value: $0) },
            headingFont: self.headingFont,
            bodyFont: self.bodyFont,
            font: self.font,
            light: self.light?.toShared(),
            dark: self.dark?.toShared()
        )
    }
}

extension PublicationTheme {
    func toiOS() -> SiteStandardLexicon.Theme.PublicationTheme {
        SiteStandardLexicon.Theme.PublicationTheme(
            type: self.type,
            backgroundColor: self.backgroundColor?.toiOS(),
            pageBackground: self.pageBackground?.toiOS(),
            primary: self.primary?.toiOS(),
            accentBackground: self.accentBackground?.toiOS(),
            accentText: self.accentText?.toiOS(),
            pageWidth: self.pageWidth.map { Int($0.intValue) },
            showPageBackground: self.showPageBackground.map { $0.boolValue },
            headingFont: self.headingFont,
            bodyFont: self.bodyFont,
            font: self.font,
            light: self.light?.toiOS(),
            dark: self.dark?.toiOS()
        )
    }
}

File: Android/app/src/main/java/uk/ewancroft/inkwell/data/model/common/SharedModelMappers.kt
package uk.ewancroft.inkwell.data.model.common

import uk.ewancroft.inkwell.data.model.atproto.BasicTheme
import uk.ewancroft.inkwell.data.model.atproto.ColorValue
import uk.ewancroft.inkwell.data.model.atproto.DocumentPreferences
import uk.ewancroft.inkwell.data.model.atproto.LegacyPalette
import uk.ewancroft.inkwell.data.model.atproto.PublicationPreferences
import uk.ewancroft.inkwell.data.model.atproto.PublicationTheme
import uk.ewancroft.inkwell.data.model.atproto.RgbColor
import uk.ewancroft.inkwell.shared.model.BasicTheme as SharedBasicTheme
import uk.ewancroft.inkwell.shared.model.BlobRef as SharedBlobRef
import uk.ewancroft.inkwell.shared.model.ColorValue as SharedColorValue
import uk.ewancroft.inkwell.shared.model.DocumentPreferences as SharedDocumentPreferences
import uk.ewancroft.inkwell.shared.model.LegacyPalette as SharedLegacyPalette
import uk.ewancroft.inkwell.shared.model.PublicationPreferences as SharedPublicationPreferences
import uk.ewancroft.inkwell.shared.model.PublicationTheme as SharedPublicationTheme
import uk.ewancroft.inkwell.shared.model.RgbColor as SharedRgbColor
import uk.ewancroft.inkwell.shared.model.StrongRef as SharedStrongRef

// ── BlobRef ────────────────────────────────────────────────────────────────

fun BlobRef.toShared(): SharedBlobRef = SharedBlobRef(
    link = link,
    size = size,
    type = type,
    mimeType = mimeType
)

fun SharedBlobRef.toAndroid(): BlobRef = BlobRef(
    link = link,
    size = size,
    type = type,
    mimeType = mimeType
)

// ── StrongRef ──────────────────────────────────────────────────────────────

fun StrongRef.toShared(): SharedStrongRef = SharedStrongRef(
    uri = uri,
    cid = cid
)

fun SharedStrongRef.toAndroid(): StrongRef = StrongRef(
    uri = uri,
    cid = cid
)

// ── Theme Types ────────────────────────────────────────────────────────────

fun RgbColor.toShared(): SharedRgbColor = SharedRgbColor(
    type = type,
    r = r,
    g = g,
    b = b
)

fun SharedRgbColor.toAndroid(): RgbColor = RgbColor(
    type = type,
    r = r,
    g = g,
    b = b
)

fun ColorValue.toShared(): SharedColorValue = SharedColorValue(
    type = type,
    r = r,
    g = g,
    b = b,
    a = a
)

fun SharedColorValue.toAndroid(): ColorValue = ColorValue(
    type = type,
    r = r,
    g = g,
    b = b,
    a = a
)

fun LegacyPalette.toShared(): SharedLegacyPalette = SharedLegacyPalette(
    background = background,
    text = text,
    accent = accent,
    link = link,
    surfaceHover = surfaceHover
)

fun SharedLegacyPalette.toAndroid(): LegacyPalette = LegacyPalette(
    background = background,
    text = text,
    accent = accent,
    link = link,
    surfaceHover = surfaceHover
)

fun BasicTheme.toShared(): SharedBasicTheme = SharedBasicTheme(
    type = type,
    background = background.toShared(),
    foreground = foreground.toShared(),
    accent = accent.toShared(),
    accentForeground = accentForeground.toShared()
)

fun SharedBasicTheme.toAndroid(): BasicTheme = BasicTheme(
    type = type,
    background = background.toAndroid(),
    foreground = foreground.toAndroid(),
    accent = accent.toAndroid(),
    accentForeground = accentForeground.toAndroid()
)

fun PublicationTheme.toShared(): SharedPublicationTheme = SharedPublicationTheme(
    type = type,
    backgroundColor = backgroundColor?.toShared(),
    pageBackground = pageBackground?.toShared(),
    primary = primary?.toShared(),
    accentBackground = accentBackground?.toShared(),
    accentText = accentText?.toShared(),
    pageWidth = pageWidth,
    showPageBackground = showPageBackground,
    headingFont = headingFont,
    bodyFont = bodyFont,
    font = font,
    light = light?.toShared(),
    dark = dark?.toShared()
)

fun SharedPublicationTheme.toAndroid(): PublicationTheme = PublicationTheme(
    type = type,
    backgroundColor = backgroundColor?.toAndroid(),
    pageBackground = pageBackground?.toAndroid(),
    primary = primary?.toAndroid(),
    accentBackground = accentBackground?.toAndroid(),
    accentText = accentText?.toAndroid(),
    pageWidth = pageWidth,
    showPageBackground = showPageBackground,
    headingFont = headingFont,
    bodyFont = bodyFont,
    font = font,
    light = light?.toAndroid(),
    dark = dark?.toAndroid()
)

// ── Preferences ────────────────────────────────────────────────────────────

fun PublicationPreferences.toShared(): SharedPublicationPreferences = SharedPublicationPreferences(
    showInDiscover = showInDiscover
)

fun SharedPublicationPreferences.toAndroid(): PublicationPreferences = PublicationPreferences(
    showInDiscover = showInDiscover
)

fun DocumentPreferences.toShared(): SharedDocumentPreferences = SharedDocumentPreferences(
    showComments = showComments,
    showMentions = showMentions,
    showRecommends = showRecommends,
    showPrevNext = showPrevNext,
    showInDiscover = showInDiscover
)

fun SharedDocumentPreferences.toAndroid(): DocumentPreferences = DocumentPreferences(
    showComments = showComments,
    showMentions = showMentions,
    showRecommends = showRecommends,
    showPrevNext = showPrevNext,
    showInDiscover = showInDiscover
)
